cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+  auto offsets = lcv.offsets();
+
+  // extract means and weights
+  auto data = lcv.parent().child(lists_column_view::child_column_index);
+  structs_column_view tdigest(data);
+  auto mean   = tdigest.child(tdigest_mean_column_index);
+  auto weight = tdigest.child(tdigest_weight_column_index);
+
+  // compute summed weights
+  auto cumulative_weights = cudf::make_fixed_width_column(data_type{type_id::FLOAT64},
+                                                          mean.size(),
+                                                          mask_state::UNALLOCATED,
+                                                          stream,
+                                                          rmm::mr::get_current_device_resource());
+  auto keys               = cudf::detail::make_counting_transform_iterator(
+    0,
+    [offsets = offsets.begin<offset_type>(), num_offsets = offsets.size()] __device__(size_type i) {
+      return static_cast<size_type>(
+               thrust::upper_bound(thrust::seq, offsets, offsets + num_offsets, i) - offsets) -
+             1;
+    });
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                keys,
+                                keys + weight.size(),
+                                weight.begin<double>(),
+                                cumulative_weights->mutable_view().begin<double>());
+
+  // output is a column of size input.size() * percentiles.size()
+  auto result = cudf::make_fixed_width_column(
+    output_type, input.size() * percentiles.size(), mask_state::UNALLOCATED, stream, mr);
+
+  constexpr size_type block_size = 256;
+  cudf::detail::grid_1d const grid(percentiles.size() * input.size(), block_size);
+  compute_percentiles_kernel<T><<<grid.num_blocks, block_size, 0, stream.value()>>>(
+    offsets.begin<offset_type>(),
+    input.size(),
+    {percentiles.begin<double>(), static_cast<size_t>(percentiles.size())},
+    mean.begin<double>(),
+    weight.begin<double>(),
+    min_col.begin<double>(),
+    max_col.begin<double>(),
+    cumulative_weights->view().begin<double>(),
+    result->mutable_view().begin<device_storage_type_t<T>>(),
+    output_type);
+
+  return result;
+}
+
+struct compute_percentiles_dispatch {
+  template <typename T>
+  std::unique_ptr<column> operator()(structs_column_view const& input,
+                                     column_view const& percentiles,
+                                     data_type output_type,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    if constexpr (cudf::is_numeric<T>() || cudf::is_fixed_point<T>()) {
+      return compute_approx_percentiles<T>(input, percentiles, output_type, stream, mr);
+    }
+    CUDF_FAIL("Invalid non-numeric output requested for percentile_approx");
+  }
+};
+
+}  // anonymous namespace
+
+void check_is_valid_tdigest_column(column_view const& col)
```

Comment:
Hm, would it be crazy to add a simple `tdigest_column_view` type that is constructible from `column_view/structs_column_view` and enforces this check? That way the interface would look like:

```
std::unique_ptr<column> percentile_approx(tdigests_column_view const& input,
                                          column_view const& percentiles,
                                          cudf::data_type output_type,
                                          rmm::mr::device_memory_resource* mr)
```

Not only would this be more descriptive/better declare intent, but it would allow you to define the struct of a tdigest column in exactly one place instead of needing to repeat it in many places (as it currently the case).

Reply:
And just to dispel any concerns about a bunch of extra code or machinery, I'm envisioning this type would be incredibly simple and just have a constructor that enforces this check and otherwise has the same interface as `structs_column_view` or `column_view`.

Reply:
Right.  I waffled on this a bit.  But I figured the use cases were so limited right now that it fell under the bar for doing it.  Suggest doing it as a followup PR in the interest of time.